bftengine/tests/bcstatetransfer/RVT_test.cpp

```diff
@@ -0,0 +1,110 @@
+// terms and conditions of the subcomponent's license, as noted in the
+// LICENSE file.
+
+#include <iostream>
+#include <string>
+#include <random>
+#include <vector>
+#include <cryptopp/integer.h>
+
+#include "gtest/gtest.h"
+#include "Logger.hpp"
+
+#define HASH_SIZE 256
+
+using namespace std;
+using namespace CryptoPP;
+
+static std::string randomString(size_t length) {
+  static auto& chrs = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
+  std::mt19937 rg{std::random_device{}()};
+  std::uniform_int_distribution<std::string::size_type> pick(1, sizeof(chrs) - 2);  // notice 0
+  std::string s;
+  s.reserve(length);
+  while (length--) s += chrs[pick(rg)];
+
+  return s;
+}
+
+struct InputValues {
+  InputValues(const string& l1, const string& l2, const string& p)
+      : leaf1(l1.c_str()), leaf2(l2.c_str()), parent(p.c_str()) {}
+  Integer leaf1;   // existing leaf val, shld size be 128 bytes?
+  Integer leaf2;   // newly added leaf val
+  Integer parent;  // mod by parent val (shld size be any different than leaf), shld be 256 byte?
+};
+
+class RVTTest : public ::testing::Test {
+ public:
+  InputValues values_{randomString(HASH_SIZE), randomString(HASH_SIZE), randomString(HASH_SIZE)};
+};
+
+TEST_F(RVTTest, basicAdditionSubtraction) {
+  auto a = values_.leaf1;
+  auto b = values_.leaf2;
+  auto c = values_.leaf1 + values_.leaf2;
+  ASSERT_EQ(a + b, c);
+  ASSERT_EQ(c - b, a);
+}
+
+TEST_F(RVTTest, cumulativeAssociativeProperty) {
+  auto a = values_.leaf1;
+  auto b = values_.leaf2;
+  auto c = values_.parent;
+  ASSERT_EQ(a + b + c, c + a + b);
+  ASSERT_NE(a - b, b - a);
+  ASSERT_EQ(a + (b + c), (c + a) + b);
+  ASSERT_NE(a + (b + c), (c + a) - b);
+}
+
+class RVTHashTestParamFixture : public RVTTest, public testing::WithParamInterface<vector<InputValues>> {};
+TEST_P(RVTHashTestParamFixture, basicSumAndModOps) {
+  auto hashes = GetParam();
+  for (auto& hash : hashes) {
+    Integer res = (hash.leaf1 + hash.leaf2) % hash.parent;
+    ASSERT_NE(res, hash.parent);
+    ASSERT_NE(res, hash.leaf1);
+    ASSERT_NE(res, hash.leaf2);
+  }
+}
+
+vector<InputValues> vec = {{randomString(HASH_SIZE), randomString(HASH_SIZE), randomString(HASH_SIZE)},
+                           {randomString(HASH_SIZE), randomString(HASH_SIZE), randomString(HASH_SIZE)}};
+INSTANTIATE_TEST_CASE_P(basicSumAndModOps, RVTHashTestParamFixture, ::testing::Values(vec), );
+
+class RVTTestParamFixture : public RVTTest, public testing::WithParamInterface<std::string> {};
+TEST_P(RVTTestParamFixture, validateRawValue) {
+  auto str = GetParam();
+  Integer input((CryptoPP::byte*)str.c_str(), str.size());
+  // std::cout << input << std::endl; // str -> int so prints accordingly
+
+  // ASSERT_NE(input, str); //type is not same
+  ASSERT_EQ(input.MinEncodedSize(), str.size());
+  // Encode to string doesn't work
+  vector<char> v_en(input.MinEncodedSize());
+  input.Encode((CryptoPP::byte*)&v_en[0], v_en.size());
```

Comment:
you should access the vector data using v_en.data()
Also, the v is hunfarian notation  - we should avoid this.
just call it encInput

Reply:
Thanks. Addressed.